projects/js-packages/number-formatters/tsconfig.json

```
@@ -0,0 +1,4 @@
+{
+	"extends": "jetpack-js-tools/tsconfig.base.json",
+	"include": [ "./index.ts", "./src" ]
+}
```

And finally for the setting things up correctly, you'll need to have this look something like
```suggestion
{
	"extends": "jetpack-js-tools/tsconfig.tsc.json",
	"include": [ "./index.ts" ],
	"compilerOptions": {
		"typeRoots": [ "./node_modules/@types/" ],
		"sourceMap": false,
		"outDir": "./build/",
		"target": "es2022",
	}
}
```
You're welcome to use whichever target is appropriate, rename "build" to "dist" if you prefer that, etc.

---

Ok. Added. Indeed I changed "build" to "dist".

The only problem is changing `"extends": "jetpack-js-tools/tsconfig.base.json"` to `"extends": "jetpack-js-tools/tsconfig.tsc.json"`. It gives the error I mentioned above

---

Ok - my replies from the editor may not be coming through. I said earlier:

with the suggested change in "extends", I get the following error in `index.ts`:

> Relative import paths need explicit file extensions in ECMAScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean './src/format-number.js'?

---

That's expected, and you should make the change it calls for.

What's going on here is that Node [requires the extension for esm imports](https://nodejs.org/docs/latest-v22.x/api/esm.html#mandatory-file-extensions), while bundlers like webpack typically don't.

`tsc` doesn't try to rewrite the imports at all, it just outputs what its given. If the tsconfig has moduleResolution set to "nodenext" rather than "bundler", it issues that error to flag that the import won't work if the output is going to be run by Node. When you're writing a library like this one, you don't know whether the project using your library is going to be using "nodenext" or "bundler", so [TypeScript recommends using the stricter setting for best compatibility](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-libraries).

---

That's interesting. Thanks for explaining. I did come across this in the long distant past. I need to get my head around it.

I don't exactly understand this part:

> When you're writing a library like this one, you don't know whether the project using your library is going to be using "nodenext" or "bundler"

How do I not know that? Does that make every package we export from wp-calypso faulty in that sense (cause this is the first time I come across this)?

Sorry for my ignorance. I'll try and sum it up in my head

---

> Does that make every package we export from wp-calypso faulty in that sense (cause this is the first time I come across this)?

Maybe? Or maybe I've misunderstood something at some point. But most likely it just hasn't come up because either all reusers are processing the packages via a bundler like webpack or they're using `tsc` with `moduleResolution` set to "bundler" (and aren't running the package via node directly either).

---

Interesting. That may well be the case. Thanks for clarifying and sharing the links above.